File: app/src/main/java/org/stepic/droid/ui/adapters/view_hoders/CourseItemViewHolder.java
package org.stepic.droid.ui.adapters.view_hoders;

import android.app.Activity;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.support.annotation.ColorInt;
import android.support.annotation.DrawableRes;
import android.support.v4.graphics.drawable.RoundedBitmapDrawable;
import android.support.v4.graphics.drawable.RoundedBitmapDrawableFactory;
import android.support.v7.widget.PopupMenu;
import android.view.HapticFeedbackConstants;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.target.BitmapImageViewTarget;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.stepic.droid.R;
import org.stepic.droid.analytic.Analytic;
import org.stepic.droid.configuration.Config;
import org.stepic.droid.core.ScreenManager;
import org.stepic.droid.core.presenters.ContinueCoursePresenter;
import org.stepic.droid.core.presenters.DroppingPresenter;
import org.stepic.droid.model.Course;
import org.stepic.droid.storage.operations.Table;
import org.stepic.droid.util.ContextMenuCourseUtil;
import org.stepic.droid.util.StepikLogicHelper;

import java.util.List;
import java.util.Locale;

import butterknife.BindColor;
import butterknife.BindView;
import butterknife.ButterKnife;

public class CourseItemViewHolder extends CourseViewHolderBase {

    @BindView(R.id.courseItemName)
    TextView courseName;

    @BindView(R.id.courseItemImage)
    ImageView courseIcon;

    @BindView(R.id.courseWidgetButton)
    TextView courseWidgetButton;

    @BindView(R.id.courseItemLearnersCount)
    TextView learnersCount;

    @BindView(R.id.courseItemMore)
    View courseItemMore;

    @ColorInt
    @BindColor(R.color.new_accent_color)
    int continueColor;

    @ColorInt
    @BindColor(R.color.join_text_color)
    int joinColor;


    private final String joinTitle;
    private final String continueTitle;
    private final DroppingPresenter droppingPresenter;
    private Drawable coursePlaceholder;
    private Table type;
    private final View itemView;
    private boolean isContinueExperimentEnabled;
    private Config config;
    private List<Course> courses;
    BitmapImageViewTarget imageViewTarget;
    private final Activity contextActivity;
    private final Analytic analytic;
    private final ContinueCoursePresenter continueCoursePresenter;
    private final ScreenManager screenManager;

    public CourseItemViewHolder(String joinTitle,
                                String continueTitle,
                                DroppingPresenter droppingCoursePresenter,
                                Drawable coursePlaceholder,
                                Table type,
                                final View itemView,
                                boolean isContinueExperimentEnabled,
                                Config config,
                                List<Course> courses,
                                final Activity contextActivity,
                                Analytic analytic,
                                ContinueCoursePresenter continueCoursePresenter,
                                ScreenManager screenManager) {
        super(itemView);
        ButterKnife.bind(this, itemView);
        this.joinTitle = joinTitle;
        this.continueTitle = continueTitle;
        this.droppingPresenter = droppingCoursePresenter;
        this.coursePlaceholder = coursePlaceholder;
        this.type = type;
        this.itemView = itemView;
        this.isContinueExperimentEnabled = isContinueExperimentEnabled;
        this.config = config;
        this.courses = courses;
        imageViewTarget = new BitmapImageViewTarget(courseIcon) {
            @Override
            protected void setResource(Bitmap resource) {
                RoundedBitmapDrawable circularBitmapDrawable =
                        RoundedBitmapDrawableFactory.create(contextActivity.getResources(), resource);
                circularBitmapDrawable.setCornerRadius(contextActivity.getResources().getDimension(R.dimen.course_image_radius));
                courseIcon.setImageDrawable(circularBitmapDrawable);
            }
        };
        this.contextActivity = contextActivity;
        this.analytic = analytic;
        this.continueCoursePresenter = continueCoursePresenter;
        this.screenManager = screenManager;
        courseWidgetButton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                int adapterPosition = getAdapterPosition();
                Course course = getCourseSafety(adapterPosition);
                if (course != null) {
                    onClickWidgetButton(course, isEnrolled(course));
                }
            }
        });
        itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onClickCourse(getAdapterPosition());
            }
        });

        itemView.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                v.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS);
                itemView.showContextMenu();
                return true;
            }
        });

        courseItemMore.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Course course = getCourseSafety(getAdapterPosition());
                if (course != null) {
                    showMore(v, course);
                }
            }
        });
    }


    private void showMore(View view, @NotNull final Course course) {
        PopupMenu morePopupMenu = new PopupMenu(contextActivity, view);
        morePopupMenu.inflate(ContextMenuCourseUtil.INSTANCE.getMenuResource(course));

        morePopupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                switch (item.getItemId()) {
                    case R.id.menu_item_info:
                        screenManager.showCourseDescription(contextActivity, course);
                        return true;
                    case R.id.menu_item_unroll:
                        droppingPresenter.dropCourse(course);
                        return true;
                    default:
                        return false;
                }
            }
        });

        morePopupMenu.show();
    }


    private void onClickCourse(int position) {
        if (position >= courses.size() || position < 0) return;
        analytic.reportEvent(Analytic.Interaction.CLICK_COURSE);
        Course course = courses.get(position);
        if (course.getEnrollment() != 0) {
            analytic.reportEvent(isContinueExperimentEnabled ? Analytic.ContinueExperiment.COURSE_NEW : Analytic.ContinueExperiment.COURSE_OLD);
            screenManager.showSections(contextActivity, course);
        } else {
            screenManager.showCourseDescription(contextActivity, course);
        }
    }


    private void onClickWidgetButton(@NotNull Course course, boolean enrolled) {
        if (enrolled) {
            analytic.reportEvent(Analytic.Interaction.CLICK_CONTINUE_COURSE);
            analytic.reportEvent(isContinueExperimentEnabled ? Analytic.ContinueExperiment.CONTINUE_NEW : Analytic.ContinueExperiment.CONTINUE_OLD);
            continueCoursePresenter.continueCourse(course); //provide position?
        } else {
            screenManager.showCourseDescription(contextActivity, course);
        }
    }

    @Nullable
    private Course getCourseSafety(int adapterPosition) {
        if (adapterPosition >= courses.size() || adapterPosition < 0) {
            return null;
        } else {
            return courses.get(adapterPosition);
        }
    }

    @Override
    public void setDataOnView(int position) {
        final Course course = courses.get(position);

        courseName.setText(course.getTitle());
        Glide
                .with(contextActivity)
                .load(StepikLogicHelper.getPathForCourseOrEmpty(course, config))
                .asBitmap()
                .placeholder(coursePlaceholder)
                .fitCenter()
                .into(imageViewTarget);

        if (course.getLearnersCount() > 0) {
            learnersCount.setText(String.format(Locale.getDefault(), "%d", course.getLearnersCount()));
            learnersCount.setVisibility(View.VISIBLE);
        } else {
            learnersCount.setVisibility(View.GONE);
        }


        if (isEnrolled(course)) {
            showContinueButton();
        } else {
            showJoinButton();
        }

        if (type == Table.enrolled) {
            courseItemMore.setVisibility(View.VISIBLE);
        } else {
            courseItemMore.setVisibility(View.GONE);
        }
    }

    private boolean isEnrolled(Course course) {
        return course.getEnrollment() != 0 && course.isActive() && course.getLastStepId() != null;
    }

    private void showJoinButton() {
        showButton(joinTitle, joinColor, R.drawable.course_widget_join_background);
    }

    private void showContinueButton() {
        showButton(continueTitle, continueColor, R.drawable.course_widget_continue_background);
    }

    private void showButton(String title, @ColorInt int textColor, @DrawableRes int background) {
        courseWidgetButton.setText(title);
        courseWidgetButton.setTextColor(textColor);
        courseWidgetButton.setBackgroundResource(background);
    }

}

File: app/src/main/java/org/stepic/droid/ui/adapters/view_hoders/CourseItemViewHolder.kt
package org.stepic.droid.ui.adapters.view_hoders

import android.app.Activity
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.support.annotation.ColorInt
import android.support.annotation.DrawableRes
import android.support.v4.graphics.drawable.RoundedBitmapDrawableFactory
import android.support.v7.widget.PopupMenu
import android.view.HapticFeedbackConstants
import android.view.View
import com.bumptech.glide.Glide
import com.bumptech.glide.request.target.BitmapImageViewTarget
import kotlinx.android.synthetic.main.new_course_item.view.*
import org.stepic.droid.R
import org.stepic.droid.analytic.Analytic
import org.stepic.droid.base.App
import org.stepic.droid.configuration.Config
import org.stepic.droid.core.ScreenManager
import org.stepic.droid.core.presenters.ContinueCoursePresenter
import org.stepic.droid.core.presenters.DroppingPresenter
import org.stepic.droid.model.Course
import org.stepic.droid.storage.operations.Table
import org.stepic.droid.util.ColorUtil
import org.stepic.droid.util.ContextMenuCourseUtil
import org.stepic.droid.util.StepikLogicHelper
import java.util.*
import javax.inject.Inject

class CourseItemViewHolder(
        val view: View,
        private val contextActivity: Activity,
        private val type: Table?,
        private val joinTitle: String,
        private val continueTitle: String,
        private val coursePlaceholder: Drawable,
        private val isContinueExperimentEnabled: Boolean,
        private val courses: List<Course>,
        private val droppingPresenter: DroppingPresenter,
        private val continueCoursePresenter: ContinueCoursePresenter) : CourseViewHolderBase(view) {

    @Inject
    lateinit var screenManager: ScreenManager

    @Inject
    lateinit var analytic: Analytic

    @Inject
    lateinit var config: Config


    private val continueColor: Int by lazy {
        ColorUtil.getColorArgb(R.color.new_accent_color, contextActivity)
    }
    private val joinColor: Int by lazy {
        ColorUtil.getColorArgb(R.color.join_text_color, contextActivity)
    }
    private var imageViewTarget: BitmapImageViewTarget

    init {
        App.component().inject(this)

        imageViewTarget = object : BitmapImageViewTarget(itemView.courseItemImage) {
            override fun setResource(resource: Bitmap) {
                val circularBitmapDrawable = RoundedBitmapDrawableFactory.create(contextActivity.resources, resource)
                circularBitmapDrawable.cornerRadius = contextActivity.resources.getDimension(R.dimen.course_image_radius)
                itemView.courseItemImage.setImageDrawable(circularBitmapDrawable)
            }
        }
        itemView.courseWidgetButton.setOnClickListener {
            val adapterPosition = adapterPosition
            val course = getCourseSafety(adapterPosition)
            if (course != null) {
                onClickWidgetButton(course, isEnrolled(course))
            }
        }
        itemView.setOnClickListener({ onClickCourse(adapterPosition) })

        itemView.setOnLongClickListener({ v ->
            v.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS)
            itemView.showContextMenu()
            true
        })

        itemView.courseItemMore.setOnClickListener { v ->
            val course = getCourseSafety(adapterPosition)
            if (course != null) {
                showMore(v, course)
            }
        }
    }


    private fun showMore(view: View, course: Course) {
        val morePopupMenu = PopupMenu(contextActivity, view)
        morePopupMenu.inflate(ContextMenuCourseUtil.getMenuResource(course))

        morePopupMenu.setOnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.menu_item_info -> {
                    screenManager.showCourseDescription(contextActivity, course)
                    true
                }
                R.id.menu_item_unroll -> {
                    droppingPresenter.dropCourse(course)
                    true
                }
                else -> false
            }
        }
        morePopupMenu.show()
    }


    private fun onClickCourse(position: Int) {
        if (position >= courses.size || position < 0) return
        analytic.reportEvent(Analytic.Interaction.CLICK_COURSE)
        val course = courses[position]
        if (course.enrollment != 0) {
            analytic.reportEvent(if (isContinueExperimentEnabled) Analytic.ContinueExperiment.COURSE_NEW else Analytic.ContinueExperiment.COURSE_OLD)
            screenManager.showSections(contextActivity, course)
        } else {
            screenManager.showCourseDescription(contextActivity, course)
        }
    }


    private fun onClickWidgetButton(course: Course, enrolled: Boolean) {
        if (enrolled) {
            analytic.reportEvent(Analytic.Interaction.CLICK_CONTINUE_COURSE)
            analytic.reportEvent(if (isContinueExperimentEnabled) Analytic.ContinueExperiment.CONTINUE_NEW else Analytic.ContinueExperiment.CONTINUE_OLD)
            continueCoursePresenter.continueCourse(course) //provide position?
        } else {
            screenManager.showCourseDescription(contextActivity, course)
        }
    }

    private fun getCourseSafety(adapterPosition: Int): Course? {
        return if (adapterPosition >= courses.size || adapterPosition < 0) {
            null
        } else {
            courses[adapterPosition]
        }
    }

    override fun setDataOnView(position: Int) {
        val course = courses[position]

        itemView.courseItemName.text = course.title
        Glide
                .with(contextActivity)
                .load(StepikLogicHelper.getPathForCourseOrEmpty(course, config))
                .asBitmap()
                .placeholder(coursePlaceholder)
                .fitCenter()
                .into(imageViewTarget)

        if (course.learnersCount > 0) {
            itemView.courseItemLearnersCount.text = String.format(Locale.getDefault(), "%d", course.learnersCount)
            itemView.courseItemLearnersCount.visibility = View.VISIBLE
        } else {
            itemView.courseItemLearnersCount.visibility = View.GONE
        }


        if (isEnrolled(course)) {
            showContinueButton()
        } else {
            showJoinButton()
        }

        itemView.courseItemMore.visibility = if (type === Table.enrolled) {
            View.VISIBLE
        } else {
            View.GONE
        }
    }

    private fun isEnrolled(course: Course?): Boolean =
            course != null && course.enrollment != 0 && course.isActive && course.lastStepId != null

    private fun showJoinButton() {
        showButton(joinTitle, joinColor, R.drawable.course_widget_join_background)
    }

    private fun showContinueButton() {
        showButton(continueTitle, continueColor, R.drawable.course_widget_continue_background)
    }

    private fun showButton(title: String, @ColorInt textColor: Int, @DrawableRes background: Int) {
        itemView.courseWidgetButton.text = title
        itemView.courseWidgetButton.setTextColor(textColor)
        itemView.courseWidgetButton.setBackgroundResource(background)
    }

}